packages/hardhat-core/src/internal/hardhat-network/provider/vm/dual.ts

```diff
@@ -89,42 +100,109 @@ export class DualModeAdapter implements VMAdapter {
   }
 
   public async getStateRoot(): Promise<Buffer> {
-    return this._ethereumJSAdapter.getStateRoot();
+    const ethereumJSRoot = await this._ethereumJSAdapter.getStateRoot();
+    const rethnetRoot = await this._rethnetAdapter.getStateRoot();
+
+    if (!ethereumJSRoot.equals(rethnetRoot)) {
+      console.trace(
+        `Different state root: ${ethereumJSRoot.toString(
+          "hex"
+        )} !== ${rethnetRoot.toString("hex")}`
+      );
+      throw new Error("Different state root");
+    }
+
+    return rethnetRoot;
   }
 
   public async getAccount(address: Address): Promise<Account> {
-    return this._ethereumJSAdapter.getAccount(address);
+    const ethereumJSAccount = await this._ethereumJSAdapter.getAccount(address);
+    const rethnetAccount = await this._rethnetAdapter.getAccount(address);
+
+    assertEqualAccounts(address, ethereumJSAccount, rethnetAccount);
+
+    return ethereumJSAccount;
   }
 
   public async getContractStorage(
     address: Address,
     key: Buffer
   ): Promise<Buffer> {
-    return this._ethereumJSAdapter.getContractStorage(address, key);
+    const ethereumJSStorageSlot =
+      await this._ethereumJSAdapter.getContractStorage(address, key);
+
+    const rethnetStorageSlot = await this._rethnetAdapter.getContractStorage(
+      address,
+      key
+    );
+
+    if (
+      !(
+        ethereumJSStorageSlot.length === 0 &&
+        rethnetStorageSlot.equals(Buffer.from([0x00]))
+      ) &&
+      !ethereumJSStorageSlot.equals(rethnetStorageSlot)
+    ) {
+      console.trace(ethereumJSStorageSlot);
+      console.trace(rethnetStorageSlot);
+      console.trace(
```

Comment:
Just FYI: `console.trace` will print the message and the full trace. For three consecutive calls, the trace is the same, so this output will be unnecessarily noisy. You can either replace all of them with `console.log` or, if you do want to see the call trace, just use one `console.trace`.